File: day20/ans.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def bstFromPreorder(self, preorder: List[int]) -> TreeNode:
        return self.rootBst(preorder, 0, len(preorder) - 1)
    
    def rootBst(self, arr, low, high):
        if low > high:
            return None
        
        if low == high:
            return TreeNode(arr[low])
        
        root = TreeNode(arr[low])
        
        right_low = None 
        for i in range(low + 1, high + 1):
            if arr[i] > arr[low]:
                right_low = i
                break
        
        if not right_low:
            root.left = self.rootBst(arr, low + 1, high)
            root.right = None
        else:
            root.left = self.rootBst(arr, low + 1, right_low - 1)
            root.right = self.rootBst(arr, right_low, high)
            
        return root
        
        

File: day20/ans.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    
    TreeNode* bstFromPreorder(vector<int>& preorder) {
        int i = 0;
        return buildTree(preorder, i, INT_MAX);
    }
    
    TreeNode* buildTree(vector<int>& preorder, int& idx, int bound) {
        if(idx == preorder.size() || preorder[idx] > bound) {
            return nullptr;
        }
        
        TreeNode* root = new TreeNode(preorder[idx]);
        idx += 1;
        
        root->left = buildTree(preorder, idx, root->val);
        root->right = buildTree(preorder, idx, bound);
        
        return root;
    }

};
